doc/manual/methods.rst

```diff
@@ -440,6 +440,198 @@ can also constrain type parameters of methods::
+
+
+2. Building a similar type with a different type parameter
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+When building generic code, there is often a need for constructing a similar
+object with some change made to the layout of the type.
+For instance, you might have some sort of abstract array with an arbitrary element type
+and want to write the computation on it with a specific element type.
+
+Because of the foible mentioned above, we must implement a method
+for each type that describes how to compute this type transform.
+There is no general transform of one subtype into another subtype with a different parameter.
+(Quick review: do you see why this is?)
+
+The subtypes of ``AbstractArray`` typically implement two methods for expressing this:
+A method to convert the input to a subtype of a specific ``AbstractArray{T, N}`` abstract type;
+and a method to make a new uninitialized matrix with a specific element type.
+Sample implementations of these can be found in the standard library.
+Here is a basic example usage of them::
+
+    input = convert(AbstractArray{Eltype}, input)
+    output = similar(Eltype, input)
+
+As an extension of this, in cases where the algorithm wants to mutate the input array,
+``convert`` is insufficient as the return value may alias the original input.
+Combining ``similar`` (to make the output array) and ``copy!`` (to fill it with the input data)
+is a generic way to express the requirement for a mutable copy of the input argument::
+
+    copy_with_eltype(Eltype, input) = copy!(similar(Eltype, input), input)
+
+
+3. Iterated dispatch
+~~~~~~~~~~~~~~~~~~~~
+
+In order to dispatch a multi-level parametric argument list,
+often it is best to separate each level of dispatch into separate functions.
+This may sound similar in approach to single-dispatch, but as we shall see below, it is still more flexible.
+
+For example, trying to dispatch on the element-type of an array will often run into ambiguous situations.
+Instead, commonly code will dispatch first on the container type,
+then recurse down to a more specific method based on eltype.
+In most cases, the algorithms lend themselves conveniently to this hierarchical approach,
+while in other cases, this rigor must be resolved manually.
+This dispatching branching can be observed, for example, in the logic to sum two matrices::
+
+    # first dispatch selects the map algorithm for element-wise summation
+    +(a::Matrix, b::Matrix) = map(+, a, b)
+    # then dispatch handles each element and selects the appropriate common element type for the computation
+    +(a, b) = +(promote(a, b)...)
+    # once the elements have the same type, they can be added via Intrinsic calls
+    +(a::T, b::T) = Core.add(a, b)
+
+
+4. Trait-based dispatch
+~~~~~~~~~~~~~~~~~~~~~~~
+
+A natural extension to the iterated dispatch above is to add a layer to
+method selection that identifies classes of types.
```

Comment:
Over in SimpleTraits.jl I write: " The main idea behind traits is to group types outside the type-hierarchy and to make dispatch work with that grouping. The difference to Union-types is that types can be added to a trait after the creation of the trait, whereas Union types are fixed after creation. The cool thing about Tim's trick is that there is no performance impact compared to using ordinary dispatch. "  Maybe an explanation along those lines could be helpful here too?